File: web/chains-202105/py/table_page.py
import json, pprint
from pathlib import Path
from aiohttp import web
from web_chains_202105 import utils

# ======================================================================

sTablePage = """<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    {stylesheets}
    {remote_scripts}
    {inline_scripts}
    <title>{table_name}</title>
  </head>
  <body>
    <h2 id="title"></h2>
{body}
  </body>
</html>
"""

# ======================================================================

def table_page(request, subtype_id, table_date):
    global sTablePage

    remote_scripts = [
        "js/jquery.js",
        "js/table-page.js",
        ]
    stylesheets = [
        "js/maps.css",
        "js/table-page.css",
        ]

    data = collect_table_data(request=request, subtype_id=subtype_id, table_date=table_date)

    inline_scripts = [
        f"table_page_data =\n{json.dumps(data, separators=(',', ':'))};",
        ]

    return web.Response(
        text=sTablePage.format(
            remote_scripts="\n    ".join(f'<script src="{script}" type="module"></script>' for script in remote_scripts),
            stylesheets="\n    ".join(f'<link rel="stylesheet" href="{stylesheet}">' for stylesheet in stylesheets),
            inline_scripts="\n    ".join(f'<script>\n{code}\n    </script>' for code in inline_scripts),
            table_name=f"{subtype_id} {table_date}",
            body="", # f"<pre>\n{pprint.pformat(data)}</pre>"
    ),
    content_type='text/html')

# ----------------------------------------------------------------------

def collect_table_data(request, subtype_id, table_date):
    from web_chains_202105.chart import get_chart

    def collect_table_data_part():
        for patt in ["i-*", "f-*", "b-*"]:
            for subdir in sorted(Path(subtype_id).glob(patt), reverse=True):
                entries = make_entries(subdir, patt[0] == "i")
                if entries:
                    yield {
                        "type": "individual" if patt[0] == "i" else "chain",
                        "chain_id": subdir.name,
                        **entries
                        }

    def make_entries(subdir, individual):
        entries = {}
        filenames = list(subdir.glob(f"*{table_date}.*"))
        if not filenames and subdir.name[0] == "b": # workaround for the backward chain step naming problem
            filenames = list(subdir.glob(f"*.{table_date}-*"))
        for filename in filenames:
            key1, key2 = keys_for_filename(filename)
            if individual and key1 == "scratch":
                key1 = "individual"
            entries.setdefault(key1, {})[key2] = str(filename)
            if key2 == "ace":
                chart = get_chart(request=request, filename=filename)
                chart_date = chart.date()
                entries[key1]["date"] = chart_date
                entries.setdefault("date", chart_date)
        return entries

    def keys_for_filename(filename):
        suffx = filename.suffixes
        if suffx[-1] == ".ace":
            if len(suffx) >= 2: # 123.20160113-20210625.incremental.ace
                return (suffx[-2][1:], "ace")
            else:
                return ("scratch", "ace")
        elif suffx[-1] == ".json":
            if len(suffx) >= 3: # 123.20160113-20210625.scratch.grid.json
                return (suffx[-3][1:], suffx[-2][1:])
            else:
                return ("scratch", suffx[-2][1:])
        else:
            return ("unknown", "".join(suffx))

    return {
        "subtype_id": subtype_id,
        "table_date": table_date,
        "parts": [part_data for part_data in collect_table_data_part() if part_data],
        **utils.format_subtype(request=request, subtype_id=subtype_id)
    }

# ----------------------------------------------------------------------


# ======================================================================

File: web/chains-202105/js/table-page.js
import * as DIR from "./directories.js";
import * as MAPS from "./maps.js";

// ----------------------------------------------------------------------

function main() {
    console.log("table_page_data", table_page_data);
    make_title();
    const parts = table_page_data.parts.reduce((accumulator, elt) => {
        accumulator[elt.type] = [...(accumulator[elt.type] || []), elt];
        return accumulator;
    }, {});
    show_parts(parts, "individual", table_page_data.subtype_id);
    show_parts(parts, "chain", table_page_data.subtype_id);
    // for (let part_data of table_page_data.parts)
    //     show_part(part_data, table_page_data.subtype_id);
    // $("body").append(`<pre>\n${JSON.stringify(table_page_data, null, 2)}\n</pre>`);
}

// ----------------------------------------------------------------------

function show_parts(parts, key, subtype_id) {
    const part_title_text = part_title(parts[key][0], subtype_id);
    if (part_title_text)
        $("body").append(`<hr>\n<h3>${part_title_text} ${parts[key][0].date} (${parts[key][0].chain_id})</h3>`);
    switch (key) {
    case "individual":
        show_individual_table_maps(parts);
        break;
    case "chain":
        show_chain_maps(parts);
        break;
    }

}

// ----------------------------------------------------------------------

function show_individual_table_maps(data, chain_data) {
    const table = $("<table></table>").appendTo("body");
    const tr_title = $("<tr class='title'></tr>").appendTo(table);
    const tr = $("<tr class='image'></tr>").appendTo(table);
    data.individual.forEach((data_individual) => {
        table_page_data.coloring.forEach((coloring, coloring_no) => {
            MAPS.td_title_append(tr_title, tr, MAPS.map_td_with_title(data_individual, "individual", coloring, coloring_no === 0));
        });
    });
    // procrustes
    data.individual.forEach((data_individual) => {
        data.chain.forEach((data_chain) => {
            for (let merge_type of ["incremental", "scratch"]) {
                MAPS.td_title_append(tr_title, tr, MAPS.pc_td_with_title(data_individual, "individual", data_chain, merge_type, table_page_data.coloring[0]));
            }
        });
    });
    // TODO: grid test
    // TODO: serum correlation widget
    // TODO: table widget
}

// ----------------------------------------------------------------------

function show_chain_maps(data) {
    data.chain.forEach((data_chain) => {
        const table = $("<table></table>").appendTo("body");
        table_page_data.coloring.forEach((coloring, coloring_no) => {
            const tr_title = $("<tr class='title'></tr>").appendTo(table);
            const tr = $("<tr class='image'></tr>").appendTo(table);
            for (let merge_type of ["incremental", "scratch"]) {
                MAPS.td_title_append(tr_title, tr, MAPS.map_td_with_title(data_chain, merge_type, coloring, coloring_no === 0));
            }
            MAPS.td_title_append(tr_title, tr, MAPS.pc_td_with_title(data_chain, "incremental", data_chain, "scratch", coloring)); // pc incremental vs. scratch
            for (let merge_type of ["incremental", "scratch"]) {
                // TODO: grid test
            }
        });
    });
}

// ----------------------------------------------------------------------

function part_title(part_data, subtype_id) {
    switch (part_data.type) {
    case "individual":
        return null;
    case "chain":
        switch (part_data.chain_id[0]) {
        case "f":
            return `<a href="chain?subtype_id=${subtype_id}&chain_id=${part_data.chain_id}" target="_blank">Chain</a>`;
        case "b":
            return `<a href="chain?subtype_id=${subtype_id}&chain_id=${part_data.chain_id}" target="_blank">Backward chain</a>`;
        }
        break;
    }
    return null;
}

// ----------------------------------------------------------------------

function make_title() {
    let prefix = `${DIR.lab_to_display[table_page_data.lab]} ${DIR.subtype_to_display[table_page_data.subtype]}`;
    if (table_page_data.subtype === "h3")
        prefix += ` ${DIR.assay_to_display[table_page_data.assay]} ${table_page_data.rbc || ""}`;
    $("#title").html(`${prefix} ${table_page_data.table_date}`);
}

// ----------------------------------------------------------------------

$(document).ready(() => main());
